### PR TITLE
add type hinting in `modules/nmap.py`

### PR DESCRIPTION
**Pull request by nnrepos**
this is a part of #2158 (hinty project).

### CODE REVIEW

File: scapy/config.py [hunk 1]
```diff
@@ -50,11 +50,13 @@
 if TYPE_CHECKING:
     # Do not import at runtime
     import scapy.as_resolvers
+    from scapy.data import NmapKnowledgeBase
     from scapy.packet import Packet
     from scapy.supersocket import SuperSocket  # noqa: F401
     import scapy.asn1.asn1
     import scapy.asn1.mib
 
+
```
Comment (p-l-) [inline scapy/config.py:59]: Please avoid unneeded change to preserve the code history as much as possible.

File: scapy/config.py [hunk 2]
```diff
@@ -765,7 +767,7 @@ class Conf(ConfClass):
     min_pkt_size = 60
     #: holds MIB direct access dictionary
     mib = None  # type: 'scapy.asn1.mib.MIBDict'
-    bufsize = 2**16
+    bufsize = 2 ** 16
```
Comment (p-l-) [inline scapy/config.py:770]: Please avoid unneeded change to preserve the code history as much as possible.

File: scapy/config.py [hunk 3]
```diff
@@ -942,12 +946,14 @@ def crypto_validator(func):
     This a decorator to be used for any method relying on the cryptography library.  # noqa: E501
     Its behaviour depends on the 'crypto_valid' attribute of the global 'conf'.
     """
+
```
Comment (p-l-) [inline scapy/config.py:949]: Please avoid unneeded change to preserve the code history as much as possible.

File: scapy/config.py [hunk 4]
```diff
@@ -942,12 +946,14 @@ def crypto_validator(func):
...
     def func_in(*args, **kwargs):
         # type: (*Any, **Any) -> Any
         if not conf.crypto_valid:
             raise ImportError("Cannot execute crypto-related method! "
                               "Please install python-cryptography v1.7 or later.")  # noqa: E501
         return func(*args, **kwargs)
+
```
Comment (p-l-) [inline scapy/config.py:956]: Please avoid unneeded change to preserve the code history as much as possible.

File: scapy/modules/nmap.py [hunk 5]
```diff
@@ -25,18 +25,17 @@
...
+from scapy.plist import SndRcvList, PacketList
 import scapy.libs.six as six
 
-
 if WINDOWS:
     conf.nmap_base = os.environ["ProgramFiles"] + "\\nmap\\nmap-os-fingerprints"  # noqa: E501
 else:
     conf.nmap_base = "/usr/share/nmap/nmap-os-fingerprints"
 
-
```
Comment (p-l-) [inline scapy/modules/nmap.py:39]: Please avoid unneeded change to preserve the code history as much as possible.

File: scapy/modules/nmap.py [hunk 6]
```diff
@@ -25,18 +25,17 @@
...
 from scapy.layers.inet import IP, TCP, UDP, ICMP, UDPerror, IPerror
-from scapy.packet import NoPayload
+from scapy.packet import NoPayload, Packet
 from scapy.sendrecv import sr
-from scapy.compat import plain_str, raw
+from scapy.compat import plain_str, raw, Dict, List, Tuple, Optional, cast, Union
+from scapy.plist import SndRcvList, PacketList
 import scapy.libs.six as six
 
-
```
Comment (p-l-) [inline scapy/modules/nmap.py:33]: Please avoid unneeded change to preserve the code history as much as possible.

File: scapy/modules/nmap.py [hunk 7]
```diff
@@ -124,12 +128,12 @@ def nmap_udppacket_sig(snd, rcv):
         res["ULEN"] = "%X" % rcv[UDPerror].len
         res["DAT"] = "E" if (
             isinstance(rcv[UDPerror].payload, NoPayload) or
-            raw(rcv[UDPerror].payload) == raw(snd[UDP].payload)
-        ) else "F"
+            raw(rcv[UDPerror].payload) == raw(snd[UDP].payload)) else "F"
```
Comment (p-l-) [inline scapy/modules/nmap.py:131]: Please avoid unneeded change to preserve the code history as much as possible.

File: scapy/modules/nmap.py [hunk 8]
```diff
@@ -196,15 +203,17 @@ def nmap_search(sigs):
 
 @conf.commands.register
 def nmap_fp(target, oport=80, cport=81):
+    # type: (str, int, int) -> Tuple[Union[int, float], List]
     """nmap fingerprinting
-nmap_fp(target, [oport=80,] [cport=81,]) -> list of best guesses with accuracy
-"""
+    nmap_fp(target, [oport=80,] [cport=81,]) -> list of best guesses with accuracy
+    """
```
Comment (p-l-) [inline scapy/modules/nmap.py:209]: Please avoid unneeded change to preserve the code history as much as possible.

**Review by p-l- — CHANGES_REQUESTED** [review]
Hi there,
Thanks for this PR! We try to maintain a code history as clean as possible, so could you please revert the changes that are not needed (e.g., whitespaces)?
Thanks!

**Comment by p-l-** [comment]
Also, you need to fix the `mypy` errors:
```
scapy/data.py:555: error: Missing type parameters for generic type "List"
scapy/config.py:53: error: Module "scapy.data" has no attribute "NmapKnowledgeBase"; maybe "KnowledgeBase"?
```

File: scapy/config.py (hunk 1 above)
Comment (NNRepos) [reply scapy/config.py:59]: ok

File: scapy/config.py (hunk 2 above)
Comment (NNRepos) [reply scapy/config.py:770]: ok

File: scapy/config.py (hunk 3 above)
Comment (NNRepos) [reply scapy/config.py:949]: ok

File: scapy/config.py (hunk 4 above)
Comment (NNRepos) [reply scapy/config.py:956]: ok

File: scapy/modules/nmap.py (hunk 5 above)
Comment (NNRepos) [reply scapy/modules/nmap.py:39]: ok

File: scapy/modules/nmap.py (hunk 6 above)
Comment (NNRepos) [reply scapy/modules/nmap.py:33]: ok

File: scapy/modules/nmap.py (hunk 7 above)
Comment (NNRepos) [reply scapy/modules/nmap.py:131]: ok

File: scapy/modules/nmap.py (hunk 8 above)
Comment (NNRepos) [reply scapy/modules/nmap.py:209]: ok

**Review by p-l- — APPROVED** [review]

**Comment by p-l-** [comment]
Thanks a lot for this contribution @nnrepos!

**Pull request merged**